src/main/java/wolox/training/repositories/BookRepository.java

```diff
@@ -1,18 +1,19 @@
 package wolox.training.repositories;
 
-import java.util.Optional;
-
 import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
-
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 import wolox.training.models.Book;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
 
     Optional<Book> findFirstByAuthorOrderByIdAsc(String author);
 
-    Iterable<Book> findAllByPublisherAndGenreAndYear(String publisher, String genere, String year);
+    @Query("SELECT u FROM Book u WHERE (:pub <= '' or u.publisher = :pub) and (:gen <= '' or"
```

Comment:
En la query en vez comparar a las distintas variables con empty se compara contra null. Se podría hacer algo como esto:

```suggestion
    @Query("SELECT u FROM Book u WHERE (:pub is null or u.publisher = :pub)"
            + " and (:gen is null or u.genre = :gen)"
            + " and (:yea is null or u.year = :yea)")
```

Reply:
Lo probe asi, pero no me funcó desde PostMan!!.. Por eso lo deje con el <=...